src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs

```diff
@@ -960,5 +1009,48 @@ void ThrowIfNotPeekLockMode()
 
             MessagingEventSource.Log.RegisterOnMessageHandlerStop(this.ClientId);
         }
+
+        public void RegisterPlugin(ServiceBusPlugin serviceBusPlugin)
+        {
+            if (serviceBusPlugin == null)
+            {
+                throw new ArgumentNullException(nameof(serviceBusPlugin), Resources.ArgumentNullOrWhiteSpace.FormatForUser(nameof(serviceBusPlugin)));
+            }
+            if (this.RegisteredPlugins == null)
+            {
+                this.RegisteredPlugins = new List<ServiceBusPlugin>();
+            }
+            else if (this.RegisteredPlugins.Any(p => p.GetType() == serviceBusPlugin.GetType()))
+            {
+                throw new ArgumentException(nameof(serviceBusPlugin), Resources.PluginAlreadyRegistered.FormatForUser(nameof(serviceBusPlugin)));
+            }
+            this.RegisteredPlugins.Add(serviceBusPlugin);
+        }
+
+        public void UnregisterPlugin(ServiceBusPlugin serviceBusPlugin)
```

Comment:
> This might be a stretch, and maybe that means its not needed for now

Then don't add it 😃 

>  I can think of a scenario where I would want to unregister when I continually receive errors with my plugin

I'm not sure about this. If you have a component that you rely on and it misbehaves, you would find the issue and redeploy the component. I might be off on this one, but feels like adding this w/o actual need would be premature.